internal/data/trace.go

```diff
@@ -0,0 +1,377 @@
+	return ITraceData{resourceSpans}
+}
+
+// SpanCount calculates the total number of spans.
+func (td ITraceData) SpanCount() int {
+	spanCount := 0
+	for _, rsl := range td.resourceSpans {
+		spanCount += len(rsl.spans)
+	}
+	return spanCount
+}
+
+// A collection of spans from a Resource.
+type ResourceSpans struct {
+	// The resource for the spans in this message.
+	// If this field is not set then no resource info is known.
+	resource *Resource
+
+	// A list of Spans that originate from a resource.
+	spans []*Span
+}
+
+func NewResourceSpans(resource *Resource, spans []*Span) *ResourceSpans {
+	return &ResourceSpans{resource, spans}
+}
+
+func (m *ResourceSpans) Resource() *Resource {
+	return m.resource
+}
+
+func (m *ResourceSpans) SetResource(r *Resource) {
+	m.resource = r
+}
+
+func (m *ResourceSpans) Spans() []*Span {
+	return m.spans
+}
+
+func (m *ResourceSpans) SetSpans(s []*Span) {
+	m.spans = s
+}
+
+type TraceID struct {
+	bytes []byte
+}
+
+func TraceIDFromBytes(bytes []byte) TraceID { return TraceID{bytes} }
+
+type SpanID struct {
+	bytes []byte
+}
+
+func SpanIDFromBytes(bytes []byte) SpanID { return SpanID{bytes} }
+
+// TraceState in w3c-trace-context format: https://www.w3.org/TR/trace-context/#tracestate-header
+type TraceState string
+
+// Span represents a single operation within a trace. Spans can be
+// nested to form a trace tree. Spans may also be linked to other spans
+// from the same or different trace and form graphs. Often, a trace
+// contains a root span that describes the end-to-end latency, and one
+// or more subspans for its sub-operations. A trace can also contain
+// multiple root spans, or none at all. Spans do not need to be
+// contiguous - there may be gaps or overlaps between spans in a trace.
+// Must use NewSpan* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type Span struct {
+	// Wrap OTLP Span.
+	orig *otlptrace.Span
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	attributes AttributesMap
+	events     []*SpanEvent
+	links      []*SpanLink
+}
+
+func NewSpan() *Span {
+	return &Span{orig: &otlptrace.Span{}}
+}
+
+// NewSpanSlice creates a slice of Spans that are correctly initialized.
+func NewSpanSlice(len int) []Span {
+	origs := make([]otlptrace.Span, len)
+	wrappers := make([]Span, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+	}
+	return wrappers
+}
+
+func (m *Span) TraceID() TraceID {
+	return TraceIDFromBytes(m.orig.TraceId)
+}
+
+func (m *Span) SpanID() SpanID {
+	return SpanIDFromBytes(m.orig.SpanId)
+}
+
+func (m *Span) TraceState() TraceState {
+	return TraceState(m.orig.Tracestate)
+}
+
+func (m *Span) ParentSpanID() SpanID {
+	return SpanIDFromBytes(m.orig.ParentSpanId)
+}
+
+func (m *Span) Name() string {
+	return m.orig.Name
+}
+
+func (m *Span) Kind() otlptrace.Span_SpanKind {
+	return m.orig.Kind
+}
```

Comment:
Similar comment here, may worth having a Kind defined in the data.

Reply:
OK. What do you think about just adding type aliases for enums that I used from the proto?

Reply:
I added a type alias. I decided not to add enums. It is pointless, we don't need them in Collector anyway. From Collector's perspective this is just a numerical status code, I just added a comment to explain the meaning.